compiler/rustc_hir_analysis/src/hir_ty_lowering/errors.rs

```diff
@@ -1215,7 +1216,111 @@ pub fn prohibit_assoc_item_binding(
+    // otherwise suggest removal of type binding
+    if let Some((def_id, segment, _)) = segment
+        && segment.args().parenthesized == hir::GenericArgsParentheses::No
+        && let hir::TypeBindingKind::Equality { term } = binding.kind
+    {
+        // Suggests removal of the offending binding
+        let suggest_removal = |e: &mut Diag<'_>| {
+            let bindings = segment.args().bindings;
+            let args = segment.args().args;
+            let binding_span = binding.span;
+
+            // Compute the span to remove based on the the position
+            // of the binding. We do that as follows:
+            //  1. Find the index of the binding in the list of bindings
+            //  2. Locate the spans preceding and following the binding.
+            //     If it's the first binding the preceding span would be
+            //     that of the last arg
+            //  3. Using this information work out whether the span
+            //     to remove will start from the end of the preceding span,
+            //     end at the start of the next span or will simply be the
+            //     span encomassing everything within the generics brackets
+
+            let Some(binding_index) = bindings.iter().position(|b| b.hir_id == binding.hir_id)
+            else {
+                bug!("a type binding exists but its HIR ID not found in generics");
+            };
+
+            let preceding_span = if binding_index > 0 {
+                Some(bindings[binding_index - 1].span)
+            } else {
+                args.last().map(|a| a.span())
+            };
+
+            let next_span = if binding_index < bindings.len() - 1 {
+                Some(bindings[binding_index + 1].span)
+            } else {
+                None
+            };
+
+            let removal_span = match (preceding_span, next_span) {
+                (Some(prec), _) => binding_span.with_lo(prec.hi()),
+                (None, Some(next)) => binding_span.with_hi(next.lo()),
+                (None, None) => {
+                    let Some(generics_span) = segment.args().span_ext() else {
+                        bug!("a type binding exists but generic span is empty");
+                    };
+
+                    generics_span
+                }
+            };
+
+            // Now emit the suggestion
+            if let Ok(suggestion) = tcx.sess.source_map().span_to_snippet(removal_span) {
+                e.span_suggestion_verbose(
+                    removal_span,
+                    "consider removing this type binding",
+                    suggestion,
+                    Applicability::MaybeIncorrect,
+                );
+            }
+        };
+
+        // Suggests replacing the type binding with normal
+        // type argument (i.e. replacing <T = A> with <T>)
+        let suggest_direct_use = |e: &mut Diag<'_>, sp: Span, is_ty: bool| {
+            if let Ok(snippet) = tcx.sess.source_map().span_to_snippet(sp) {
```

Comment:
I tested with the following macro code:
```rust
macro_rules! use_directly_test {
    ($x:ident, $y:ty) => {
        pub trait MyTr1<$x> {
        }

        impl MyTr1<$x = $y> for usize {
        }
    };
}

pub fn main() {
    use_directly_test!(MyArg, usize);
}
```

Using the `span_to_snippet` approach (i.e. the currently implemented approach) produces the following diagnostics:
```
error[E0229]: associated type bindings are not allowed here
  --> D:/Projects/rust_dev/rustc/122162-impl-type-binding-suggestion.rs:31:20
   |
31 |         impl MyTr1<$x = $y> for usize {
   |                    ^^^^^^^ associated type not allowed here
...
50 |     use_directly_test!(MyArg, usize);
   |     -------------------------------- in this macro invocation
   |
   = note: this error originates in the macro `use_directly_test` (in Nightly builds, run with -Z macro-backtrace for more info)
help: to use `usize` as a generic argument specify it directly
   |
31 |         impl MyTr1<usize> for usize {
   |                    ~~~~~
```

whereas using the approach where we suggest removal of `T = ` / `C = ` produces the following:
```
error[E0229]: associated type bindings are not allowed here
  --> D:/Projects/rust_dev/rustc/122162-impl-type-binding-suggestion.rs:31:20
   |
31 |         impl MyTr1<$x = $y> for usize {
   |                    ^^^^^^^ associated type not allowed here
...
50 |     use_directly_test!(MyArg, usize);
   |     -------------------------------- in this macro invocation
   |
   = note: this error originates in the macro `use_directly_test` (in Nightly builds, run with -Z macro-backtrace for more info)
help: to use `usize` as a generic argument specify it directly
   |
50 -     use_directly_test!(MyArg, usize);
50 +     use_directly_test!(usize);
   |
```

Neither set of diagnostics is great as they both point to values passed to the macro rather than any code inside the macro. However, the second set of diagnostics is IMHO less clear and more confusing than the first. So I feel we should stick with the currently implemented approach. What do you think?

Reply:
Thanks a lot for not blindly following my advice!

Right, the second output is a no-go! We generally shouldn't tamper with the token stream of a macro call since that's almost never what we want and can lead to more errors.

The first (current) output is also not ideal as you've already mentioned since that's very likely not what the user wants either (`usize` is specific to the call-site and shouldn't be moved to the def-site). It's possible that `use_directly_test` isn't even defined the current crate in which case modification is impossible.

Reply:
The way we can combat this is by using one or more tools that are at our disposal: `Span::can_be_used_for_suggestions` (to suppress bad suggestions), `Span::eq_ctxt` (to check if two things live in different syntax contexts) and methods like `Span::find_ancestor_inside` (to follow / resolve a chain of macro expansions).

Reply:
Where this gets a bit “grindy” / time-consuming is checking various combinations of metavars / non-metavars / macro calls to ensure everything works correctly. So `Tr<Lhs = $rhs>`, `Tr<$Lhs = rhs>`, `Tr<$Lhs = $rhs>` (which you already did), `Tr<Lhs = Call!()>`, `Tr<$Lhs = Call!()>`, `Tr $Args` (not sure if that's gonna work), `Tr<$Binding>`, etc.

Reply:
If you have the time and energy, please check if some of these things lead to really bad output and if so if we can leverage `can_be_used_for_suggestions`, `eq_ctxt`, `find_ancestor_inside` to remedy that.

Note that otherwise this PR is good to go! Also note that a lot of diagnostic code in rustc doesn't properly deal with metavars (`$var`) and macro calls, so it's okay if it's not perfect. It can be fixed in follow-up PRs, too, by you or by me whatever works out best.

Reply:
Same for the `suggest_removal` case!

Reply:
Thanks @fmease . What you said makes a lot of sense (and is quite educative for me). 

I may not be available for the next few days so I'm inclined to let this PR be merged and take up these changes later as you suggested.